### PR TITLE
Update to DBConnection 2.0

### PR DESCRIPTION
**Pull request by josevalim**
Includes #196 and #234. Note DBConnection 2.0 has not been released yet. There is also some work pending:

 * [x] Use database transaction status to prevent transaction errors
 * [x] Support new low level cursor API with per fetch max rows
 * [ ] Handle execute errors while using cursors
 * [ ] Test that transaction status is updated for more query result packets

/cc @wojtekmach @fishcakez

### CODE REVIEW

File: test/start_test.exs [hunk 1]
```diff
@@ -1,14 +1,25 @@
...
+
+    assert capture_log(fn ->
+      {:ok, pid} = Mariaex.start_link([username: "mariaex_user", password: "mariaex_pass", database: "non_existing"] ++ opts)
+      assert_receive {:EXIT, ^pid, :killed}, 5000
+    end) =~ "** (Mariaex.Error) (1049): Unknown database 'non_existing'"
+
+    assert capture_log(fn ->
+      {:ok, pid} = Mariaex.start_link([username: "non_existing", database: "mariaex_test"] ++ opts)
+      assert_receive {:EXIT, ^pid, :killed}, 5000
+    end) =~ "** (Mariaex.Error) (1045): Access denied for user 'non_existing'@'localhost'"
```
Comment (fishcakez) [inline test/start_test.exs]: We should check up to and including the `@` I guess. Unsure what happens on other languages though. Ideally we would have similar style to postgrex where every error code has a matching atom.

File: test/transaction_test.exs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    assert DBConnection.status(conn, opts) == :transaction
+    DBConnection.rollback!(conn, opts)
+    assert DBConnection.status(pid, opts) == :transaction
+
+   DBConnection.commit!(pid, [mode: :transaction])
+   assert DBConnection.status(pid) == :idle
+   assert query("SELECT 42", []) == [[42]]
+  end
+
+  test "can not begin, commit or rollback savepoint transaction if not begun", context do
```
Comment (fishcakez) [inline test/transaction_test.exs]: we should mention savepoint in this name as its confusing to say can not begin if mnot not begun without that context.

File: lib/mariaex/protocol.ex [hunk 3]
```diff
@@ -668,22 +664,45 @@ defmodule Mariaex.Protocol do
...
 
-  defp handle_ok_packet(packet(msg: ok_resp(affected_rows: affected_rows, last_insert_id: last_insert_id)), _query, s) do
+  defp handle_ok_packet(packet(msg: ok_resp(affected_rows: affected_rows, last_insert_id: last_insert_id, status_flags: flags)), _query, s) do
     result = %Mariaex.Result{columns: [], rows: nil, num_rows: affected_rows,
       last_insert_id: last_insert_id, connection_id: s.connection_id}
-    {:ok, {result, nil}, clean_state(s)}
+    {:ok, {result, nil}, clean_state(s, flags)}
+  end
+
+  defp clean_state(state, flags) do
```
Comment (fishcakez) [inline lib/mariaex/protocol.ex]: We need to test that we always call this function after a query finishes.

**Comment by fishcakez** [comment]
> Handle execute errors while using cursors

Need some time to think about that this TODO means. Likely on the first `handle_next` call we need to test all code paths for cursors, and currently don't.

File: test/transaction_test.exs (hunk 2 above)
Comment (fishcakez) [reply test/transaction_test.exs]: You're right, this test is wonderful ;)

**Comment by fishcakez** [comment]
This looks good to merge if we open issues for the TODO items so we can block ecto dev.

File: lib/mariaex.ex [hunk 4]
```diff
@@ -237,16 +236,9 @@ defmodule Mariaex do
       Mariaex.execute(conn, query, ["%my%"])
   """
   @spec execute(conn, Mariaex.Query.t, list, Keyword.t) ::
-    {:ok, Mariaex.Result.t} | {:error, Mariaex.Error.t}
+    {:ok, Mariaex.Query.t, Mariaex.Result.t} | {:error, Mariaex.Error.t}
```
Comment (fishcakez) [inline lib/mariaex.ex]: I think this can now return other exceptions in the error case?

File: lib/mariaex.ex [hunk 5]
```diff
@@ -291,10 +278,8 @@ defmodule Mariaex do
     case DBConnection.close(conn, query, defaults(opts)) do
       {:ok, _} ->
         :ok
-      {:error, %Mariaex.Error{}} = error ->
-        error
       {:error, err} ->
-        raise err
+        err
```
Comment (fishcakez) [inline lib/mariaex.ex]: I dont think this right, don't we want to return the error tuple?

File: lib/mariaex.ex [hunk 6]
```diff
@@ -419,25 +404,21 @@ defmodule Mariaex do
...
 
   defp prepare_binary(conn, query, opts) do
     case DBConnection.prepare(conn, query, defaults(opts)) do
       {:ok, _} = ok ->
         ok
-      {:error, %Mariaex.Error{}} = error ->
-        error
       {:error, err} ->
-        raise err
+        err
```
Comment (fishcakez) [inline lib/mariaex.ex]: Same here.

**Comment by josevalim** [comment]
This is ready to go. I have opened issue #238 to track what is left. We should close #196 after merging this.

**Comment by fishcakez** [comment]
Thank you :heart:

**Pull request merged**